walker.go

```diff
@@ -147,23 +149,78 @@ func Walk(ctx context.Context, p string, opt *WalkOpt, fn filepath.WalkFunc) err
 		}
 
 	passedFilter:
-		stat, err := mkstat(origpath, path, fi, seenFiles)
-		if err != nil {
-			return err
+		// don't report partial files yet, we might later if we eventually find a
+		// non-partial file in this path
+		if partial && fi.IsDir() {
+			return nil
 		}
 
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
-			if opt != nil && opt.Map != nil {
-				if allowed := opt.Map(stat.Path, stat); !allowed {
-					return nil
+		type pathInfo struct {
+			OrigPath string
+			Path     string
+			Info     os.FileInfo
+		}
+
+		// collect any parent path that have not been previously reported
+		// so we can report it first
+		newPaths := []pathInfo{}
+		origDir := filepath.Dir(origpath)
+		if _, ok := seenDirs[origDir]; !ok {
+			relDir, err := filepath.Rel(root, origDir)
+			if err != nil {
+				return err
+			}
+			dirParts := strings.Split(relDir, string(filepath.Separator))
+			for i := 0; i < len(dirParts); i++ {
+				dir := filepath.Join(dirParts[0 : i+1]...)
+				if dir == "." {
+					continue
+				}
+				origDir := filepath.Join(root, dir)
+				if _, ok := seenDirs[origDir]; !ok {
+					fi, err := os.Stat(origDir)
```

Comment:
What is this stat call for?

Reply:
The `fn` called later requires an os.FileInfo.  The updated code will skip the `fn` call the first time we see a `partial` matched directory, so we lose the original stat results.  Later when we determine the directory actually needs to exist (via non partial match in subdirectory) then we re-stat the original directory so we can pass the correct os.FileInfo back to the `fn` callback.  An alternate approach would be to capture the original stat calls in a map to reuse in the case we find a non-partial match, but I was concerned about the memory cost of preserving all the stats for a large directory tree and figured the cost of res-stating the directory was a better trade off.

Reply:
I overlooked that `fi` was used and only saw the error handling (that I thought was already handled before).

Reply:
>  but I was concerned about the memory cost of preserving all the stats for a large directory tree

I guess you only need to keep a stack of the current parents and can reuse the space once walking one directory has been completed.

Reply:
I will look into that, it seems plausible that we can just keep a small stack of FileInfo and then pop/push as we traverse the tree.